syntax = "proto3";

// Google Chat protobuf definitions (minimal subset for API requests)

message RequestHeader {
  int32 client_type = 2;  // 3 = WEB
  int64 client_version = 3;
  string locale = 4;      // "en"
  ClientFeatureCapabilities client_feature_capabilities = 5;
}

message ClientFeatureCapabilities {
  bool enable_all_features = 1;
  int32 spaces_level_for_testing = 2;
  int32 dms_level_for_testing = 3;
  int32 post_rooms_level = 4;
  int32 spam_room_invites_level = 5;  // 2 = FULLY_SUPPORTED
  int32 tombstone_level = 6;
  int32 rich_text_viewing_level = 7;
}

message SpaceId {
  string space_id = 1;
}

message DmId {
  string dm_id = 1;
}

enum UserType {
  HUMAN = 0;
  BOT = 1;
}

message UserId {
  string id = 1;
  UserType type = 2;
}

message MemberId {
  UserId user_id = 1;
}

message MembershipId {
  MemberId member_id = 1;
  SpaceId space_id = 2;
  GroupId group_id = 3;
}

message GroupId {
  SpaceId space_id = 1;
  DmId dm_id = 3;
}

message TopicId {
  GroupId group_id = 1;
  string topic_id = 2;
}

message MessageParentId {
  TopicId topic_id = 1;
}

message ReferenceRevision {
  int64 timestamp = 1;
}

// Request messages

enum WorldSectionType {
  WORLD_SECTION_TYPE_UNSPECIFIED = 0;
  STARRED_DIRECT_MESSAGE_PEOPLE = 1;
  STARRED_ROOMS = 2;
  STARRED_DIRECT_MESSAGE_BOTS = 3;
  NON_STARRED_DIRECT_MESSAGE_PEOPLE = 4;
  NON_STARRED_ROOMS = 5;
  NON_STARRED_DIRECT_MESSAGE_BOTS = 6;
  ALL_DIRECT_MESSAGE_PEOPLE = 7;
  ALL_ROOMS = 8;
  ALL_DIRECT_MESSAGE_BOTS = 9;
  INVITED_DM_PEOPLE = 10;
  SPAM_INVITED_DM_PEOPLE = 11;
  STARRED_DIRECT_MESSAGE_EVERYONE = 12;
  NON_STARRED_DIRECT_MESSAGE_EVERYONE = 13;
  ALL_DIRECT_MESSAGE_EVERYONE = 14;
  STARRED_DMS_AND_STARRED_UNNAMED_ROOMS = 15;
  NON_STARRED_DMS_AND_NON_STARRED_UNNAMED_ROOMS = 16;
}

message WorldSection {
  WorldSectionType world_section_type = 1;
}

message WorldFilter {}

message WorldSectionRequest {
  int32 page_size = 1;
  WorldSection world_section = 2;
  int64 anchor_sort_timestamp_micros = 3;
  WorldFilter world_filter = 4;
  int32 num_world_items_with_snippet = 5;
  string pagination_token = 6;
}

message PaginatedWorldRequest {
  enum FetchOptions {
    UNKNOWN = 0;
    EXCLUDE_GROUP_LITE = 1;
    FETCH_BOTS_IN_HUMAN_DM = 2;
    FETCH_SPACE_INTEGRATION_PAYLOADS = 3;
    FETCH_GROUPS_D3_POLICIES = 4;
  }

  RequestHeader request_header = 1;
  repeated WorldSectionRequest world_section_requests = 2;
  string world_consistency_token = 3;
  repeated FetchOptions fetch_options = 4;
  bool fetch_from_user_spaces = 5;
  bool receive_world_update_notifications = 6;
  bool fetch_snippets_for_unnamed_rooms = 7;
}

message ListTopicsRequest {
  RequestHeader request_header = 1;
  int32 page_size_for_topics = 2;
  int32 page_size_for_replies = 3;
  int32 page_size_for_unread_replies = 6;
  int32 page_size_for_read_replies = 7;
  GroupId group_id = 8;
  ReferenceRevision group_not_older_than = 9;
}

message ListMessagesRequest {
  RequestHeader request_header = 1;
  MessageParentId parent_id = 2;
  int32 page_size = 3;
}

message GetGroupRequest {
  RequestHeader request_header = 1;
  GroupId group_id = 2;
}

message GetSelfUserStatusRequest {
  RequestHeader request_header = 1;
}

message GetMembersRequest {
  repeated MemberId member_ids = 1;
  repeated MembershipId membership_ids = 2;
  RequestHeader request_header = 100;
}
